### PR TITLE
Reimplement style values atop NSExpression

### LINKED ISSUE #8074: Reimplement MGLStyleValue atop NSExpression
MGLStyleValue should be reimplemented as a subclass of or wrapper around NSExpression, and MGLStyleLayer layout and paint properties should accept NSExpressions instead of MGLStyleValues.

Especially once we introduce support for arbitrary arithmetic and conditional expressions in style values (mapbox/mapbox-gl-js#4077 and possibly mapbox/mapbox-gl-js#4079), maintaining a custom vocabulary and type system around these values will become a major burden. Though it’s nice that the current MGLStyleValue system is strongly typed, taking better advantage of type inference in Swift, our API surface area and the learning curve will increase quite a bit as we essentially try to reinvent a statically-typed NSExpression.

NSExpression performs runtime type checking for initialization and static type checking for inspection. For maximum developer comfort, it features a format string mini-language and a series of strongly-typed getters that are applicable to certain kinds of expressions. MGLVectorStyleLayer takes advantage of all these features by representing filters as NSPredicates, which consist of NSExpressions. As with NSPredicate, developers are already familiar with NSExpression and its format string syntax, so there won’t be much of a learning curve at all.

The one thing NSExpression doesn’t provide is the concept of a style function. It does provide a rather awkward extensibility mechanism; we could extend NSExpression to understand functions and provide a more ergonomic, strongly typed API for working with functions. For constant values and expressions, the developer would use NSExpression directly. https://github.com/mapbox/mapbox-gl-native/issues/5970#issuecomment-253087889 has a sketch of the proposed API.

My understanding is that mapbox/mapbox-gl-js#4077 would require a style specification version bump, which would precipitate an SDK major version bump, so we probably don’t need to worry so much about backwards compatibility for this refactoring.

/cc @boundsj @incanus @lucaswoj @jfirebaugh

### LINKED ISSUE #9948: Split out guide about predicates
[`MGLVectorStyleLayer.predicate`](https://www.mapbox.com/ios-sdk/api/3.6.2/Classes/MGLVectorStyleLayer.html#/c:objc%28cs%29MGLVectorStyleLayer%28py%29predicate) has documentation with enough detail that it could stand on its own as a jazzy guide, similar to the one for [tile URL templates](https://www.mapbox.com/ios-sdk/api/3.6.2/tile-url-templates.html). We already use predicates in multiple parts of the API, including the feature querying methods of MGLMapView and MGLVectorSource, so centralizing the documentation in a guide would make it easier to see all those APIs as being related. Once we replace MGLStyleValue with NSExpression (#8074), it’ll be all the more important to document expressions and predicates in one place.

This documentation doesn’t differ at all between iOS and macOS, so all we need is a Markdown document in platform/darwin/docs/guides/. However, we’ll also need a small amount of code generation if we introduce any code examples into that guide.

The new guide should link to [this section](https://www.mapbox.com/ios-sdk/api/3.6.2/for-style-authors.html#filtering-sources) of the “Information for Style Authors” guide that discusses predicates. I think that section can stay where it is, because folks who only work with the runtime styling API won’t need to see it.

/cc @captainbarbosa @jmkiley @anandthakker

### PR DESCRIPTION
**Pull request by 1ec5**
This PR migrates the entire iOS/macOS style layer API – all the layout and paint properties – from the MGLStyleValue class cluster to the NSExpression class provided by Foundation, backed by @anandthakker’s work on `mbgl::style::expression::Expression` in #9439. This is a major, backwards-incompatible change to the iOS and macOS map SDKs’ public APIs. In the case of the iOS map SDK, it’ll require a bump to version 4️⃣. 0️⃣. 0️⃣. 🙊 Since this will be a major version bump, I don’t expect us to leave the older MGLStyleValue-based API around in any form, except perhaps a migration guide.

In exchange for all the upheaval, developers will be able to set individual attributes to more complex values incorporating conditionals and arithmetic while reducing the amount of code and eliminating the longwinded class names. See the detailed rationale for this overhaul in #8074.

For example, this screenshot shows local-language labels with glosses in the system language – but only where needed. Notice how Helsinki is labeled simply “Helsinki”.

labels

## Contents

This PR consists of documentation improvements, rote changes to parameter and return types all over the place, and three kinds of conversions:

1. A conversion from NSExpression to JSON-compatible Foundation types like NSString and NSNumber (`NSExpression.mgl_jsonExpressionObject`)
2. A conversion from those JSON-compatible types back to NSExpression (`+[NSExpression mgl_expressionWithJSONObject:]`)
3. A conversion from `mbgl::style::expression::Expression` to the JSON-compatible types (`MGLJSONObjectFromMBGLExpression()`)

The three conversions are quite loosely coupled, which raises the risk of undefined behavior and poor round-tripping. The existing tests do a decent job of testing round-tripping for correct input, but not for degenerate cases. I’m eagerly awaiting #10714, which will allow us to ditch the third conversion in favor of a formal method that converts `mbgl::style::expression::Expression` to a string or JSON-like structure. In the meantime, I’ve added a few things to core to make the expression internals more accessible to SDK code. In driving a beeline path to a running SDK, I probably glossed over some memory management details.

## Design considerations

NSExpression is a mixed bag: on the one hand, it’s familiar to Objective-C and Swift developers, being used extensively in Core Data and NSArray filtering. Developers unfamiliar with this class can find quite a bit of information on it online, most of it applicable to style attributes. On the other hand, the style specification’s expression model goes beyond NSExpression in many ways. NSExpression’s format string parser is rather inflexible, perhaps for interoperability reasons.

At various points, I considered rolling a custom Yacc or Bison parser for expression format strings. However, I think NSExpression just barely meets our needs. We’ll need to work closely with the GL JS project to ensure no fancy syntaxes emerge that become impossible to represent in NSExpression. Until we release v4.0.0, cutting over to a Bison-generated NSExpression workalike remains a possibility.

NSExpression comes in two flavors: if you use specialized initializers like `+[NSExpression expressionForFunction:selectorName:arguments:]`, you get a modicum of type checking at the cost of brevity. By contrast, if you use the format string initializer, you shift nearly all type checking to runtime. The format string syntax is intuitive for the most common operations (in terms of general coding usage) but becomes bewildering for less common operations. Overall, once the convenience initializers are implemented, I think we’ll end up striking a sensible balance between type safety and readability, at least from the perspective of an Objective-C programmer.

The screenshot above is the result of applying this expression to the `text` property of various MGLSymbolStyleLayers:

```objc
layer.text = [NSExpression expressionWithFormat:@"TERNARY(name = name_en, name, FUNCTION(name, 'stringByAppendingString:', '\\n(', name_en, ')'))"];
```

Alternatively:

```objc
    NSExpression *name = [NSExpression expressionForKeyPath:@"name"];
    NSExpression *englishName = [NSExpression expressionForKeyPath:@"name_en"];
    NSPredicate *conditional = [NSComparisonPredicate predicateWithLeftExpression:name
                                                                  rightExpression:englishName
                                                                         modifier:NSDirectPredicateModifier
                                                                             type:NSEqualToPredicateOperatorType
                                                                          options:0];
    NSExpression *nameWithGloss = [NSExpression expressionForFunction:name
                                                         selectorName:@"stringByAppendingString:"
                                                            arguments:@[@"\\n(", englishName, @")"]];
    layer.text = [NSExpression expressionForConditional:conditional
                                         trueExpression:name
                                        falseExpression:nameWithGloss];
```

## Progress

Some expression operators defined in the style specification are still unimplemented. I prioritized operators used in current, function-based styles, followed by operators with NSExpression equivalents. So a style that has already been migrated to expressions could break some of the getters. Conversely, some NSExpression functions, such as `mode:` and `stddev:`, can’t be implemented generically without built-in support in mbgl and the style specification. In lieu of an inline checklist, please see the enclosed “[Predicates and Expressions](https://github.com/mapbox/mapbox-gl-native/blob/1ec5-e8n-8074/platform/darwin/docs/guides/Predicates%20and%20Expressions.md)” guide, as well as the massive table under “[Expression operators](https://github.com/mapbox/mapbox-gl-native/blob/1ec5-e8n-8074/platform/ios/docs/guides/For%20Style%20Authors.md#expression-operators)” in the “Information for Style Authors” guide. The “[Using Style Functions at Runtime](https://github.com/mapbox/mapbox-gl-native/blob/1ec5-e8n-8074/platform/ios/docs/guides/Using%20Style%20Functions%20at%20Runtime.md)” guide, soon to be renamed, illustrates what each kind of legacy style function looks like after a migration to expressions.

This has been a long road, and there’s quite a bit more to come:

* [x] Design NSExpression representation of style specification expressions
    * [x] Operators without native NSExpression equivalents
    * [x] String concatenation
    * [x] Interpolation
    * [x] Variable assignment
    * [ ] ~~Type assertions~~ – tail work, see #11004
* [x] Convert NSExpression to JSON objects
    * [x] Convert every operator that has an analogue in NSExpression
    * [x] Convert enough operators to represent all the layer properties in Mapbox Streets v10
    * [x] Convert enough operators to replace examples in the style function guide
    * [ ] ~~Convert every operator in the style specification~~ – tail work, see #11004, #11006, #11009, #11010, #11011, #11012
    * [x] Synthesize some NSExpression operators that have no equivalent in the style specification
    * [ ] ~~Synthesize some NSPredicate operators that have no equivalent in the style specification~~ – tail work, see #11013
    * [ ] ~~Implement convenience methods for convoluted expression syntaxes~~ – tail work, see #11007
    * [x] Accept literal, aggregate, and constant value expressions interchangeably as function arguments
    * [x] Raise exceptions for edge cases instead of failing gracefully
* [x] Convert JSON objects to NSExpression
    * [x] Convert every operator that has an analogue in NSExpression
    * [x] Convert enough operators to represent all the layer properties in Mapbox Streets v10
    * [x] Convert enough operators to replace examples in the style function guide
    * [ ] ~~Convert every operator in the style specification~~ – tail work
    * [x] Expose this conversion publicly for React Native SDK
* [x] Convert `mbgl::style::expression::Expression` to JSON objects
    * [x] Implement accessors on mbgl expression classes
    * [x] Convert every operator that has an analogue in NSExpression
    * [x] Convert enough operators to represent all the layer properties in Mapbox Streets v10
    * [x] Convert enough operators to replace examples in the style function guide
    * [ ] ~~Convert every operator in the style specification~~ – tail work
    * [x] Raise exceptions for edge cases instead of failing gracefully
    * [ ] ~~Push this conversion down into mbgl for #10714~~ – tail work
* [x] Expand expression coverage to the whole runtime styling API
    * [x] Implement `MGLSymbolStyleLayer.text` setter
    * [x] Implement `MGLSymbolStyleLayer.text` getter
    * [x] Replace style values with expression in code generation templates
    * [x] Fix build errors on setters and getters that are more complex than `text`
    * [x] Implement light properties – tail work
    * [x] Migrate `MGLStyle.localizesLabels` to modify constant expressions
    * [ ] ~~Migrate `MGLStyle.localizesLabels` to modify interpolated expressions~~ (#10713, tail work, blocked by #10714 and/or #10944)
* [x] Migrate example applications to expressions
    * [x] Migrate macosapp style options
    * [x] Migrate iosapp style options
* [x] Improve test coverage
    * [x] Test conversion from NSExpression to JSON
    * [x] Test conversion from JSON to NSExpression
    * [x] Test native evaluation of custom functions using NSExpression
    * [x] Test conversion from `mbgl::style::expression::Expression` to JSON
* [x] Write documentation
    * [x] Function/operator support matrix
    * [x] Factor out predicate overview guide (fixes #9948)
    * [x] Expressions overview guide
    * [x] Method-level documentation
    * [x] Example code in headers
    * [ ] ~~Line-level documentation~~ – tail work
    * [x] Update changelogs
    * [ ] ~~Style value migration guide~~ – tail work
    * [ ] ~~Rework style function overview guide~~ – tail work
    * [x] Update expressions in style function overview guide to account for #10861
* [x] Get code reviews
    * [x] From a lucky member of the iOS team 🍀
    * [x] From a lucky member of the core team 🎰
    * [ ] ~~From a lucky member of the support team 🎲~~
* [x] Clean up code
    * [x] Remove MGLStyleValue and associated code
    * [ ] ~~Derive expression documentation from style specification JSON~~ – tail work
    * [x] Reorder core commits first
    * [x] Remove heatmap-specific content
    * [x] Remove to-do warnings and file issues for tail work

Fixes #8074 and fixes #9948.

<details>
<summary>Bloopers</summary>
brackets
</details>

/cc @mapbox/ios @anandthakker @jfirebaugh @ivovandongen @lucaswoj @nitaliano

### CODE REVIEW

File: include/mbgl/style/function/camera_function.hpp [hunk 1]
```diff
@@ -74,6 +74,8 @@ class CameraFunction {
 
     bool useIntegerZoom = false;
     
+    const expression::Expression* getExpression() const { return expression.get(); }
```
Comment (jfirebaugh) [inline include/mbgl/style/function/camera_function.hpp]:
```
const expression::Expression& getExpression() const { return *expression; }
```

(x3)

Comment (1ec5) [reply include/mbgl/style/function/camera_function.hpp]: 👍 Done.

**Comment by 1ec5** [comment]
One of the things I’d like to improve ahead of the final release is simplying the syntax for common operations like string concatenation and interpolation.

One possibility would be to somehow categorize [the private `_NSPredicateUtilities` class](https://github.com/nst/iOS-Runtime-Headers/blob/master/Frameworks/Foundation.framework/_NSPredicateUtilities.h) that determines the set of built-in functions. I’m not sure how feasible that would be, but the result would be fairly elegant:

```swift
NSExpression(format: "mgl_concat({'foo', 'bar', 'baz'})")
```

Another possibility would be overload existing operators in ways that can be parsed but not evaluated by NSExpression. For example, `[NSExpression expressionWithFormat:@"'foo' & 'bar'"]` would be an elegant, spreadsheet-like way of concatenating two strings. This format gets parsed into a `bitwiseAnd:with:` function with the two strings as arguments. Naturally, calling `-expressionValueWithObject:context:` raises an exception, because strings are invalid bitwise operator arguments. Getting `-expressionValueWithObject:context:` to work is a low priority, because not all of the custom functions can be natively evaluated either.

The most straightforward option would be to categorize NSExpression itself with some convenience methods like `-expressionForMGLFunction:selectorName:arguments:` and `-mgl_expressionByAppendingStrings:`. These methods would sugarcoat the ugly `FUNCTION()` syntax. Using them in conjunction with a format string would still be inconvenient:

```swift
NSExpression(format: "TERNARY(ele > 0, %@, name)",
             NSExpression(forMGLFunction: NSExpression(forKeyPath: "name"),
                          selectorName: "stringByAppendingStrings:",
                          arguments: [" (",
                                      NSExpression(forKeyPath: "ele"),
                                      ")"]))
```

or:

```swift
NSExpression(format: "TERNARY(ele > 0, %@, name)",
             NSExpression(forKeyPath: "name").mgl_appendingStrings(" (",
                                                                   NSExpression(forKeyPath: "ele"),
                                                                   ")")
```

**Comment by 1ec5** [comment]
Continuing the discussion about operator conveniences in https://github.com/mapbox/mapbox-gl-native/issues/8074#issuecomment-355383677.

File: platform/darwin/test/MGLDocumentationGuideTests.swift [hunk 2]
```diff
@@ -175,28 +166,17 @@ class MGLDocumentationGuideTests: XCTestCase, MGLMapViewDelegate {
...
-                                              sourceStops: categoricalStops,
-                                              attributeName: "type",
-                                              options: [.defaultValue: MGLStyleValue<UIColor>(rawValue: .blue)])
+            let colors: [UIColor] = [.orange, .red, .yellow, .blue]
         #endif
+        
+        layer.circleColor = NSExpression(format:
+            "TERNARY(FUNCTION(type, 'stringValue') = 'earthquake', %@, " +
+            "TERNARY(FUNCTION(type, 'stringValue') = 'explosion', %@, " +
+            "TERNARY(FUNCTION(type, 'stringValue') = 'quarry blast', %@, " +
```
Comment (1ec5) [inline platform/darwin/test/MGLDocumentationGuideTests.swift]: The calls to `stringValue` will go away with #10861.

/cc @jmkiley

File: platform/darwin/src/MGLLight.h [hunk 3]
```diff
@@ -117,15 +127,19 @@ MGL_EXPORT
...
+ * Conditional expressions
+ * Variable assignments and references to assigned variables
+ * Interpolation and step functions applied to the `$zoomLevel` variable
  
- * `MGLConstantStyleValue`
- * `MGLCameraStyleFunction` with an interpolation mode of:
-   * `MGLInterpolationModeExponential`
-   * `MGLInterpolationModeInterval`
+ This property does not support applying interpolation or step functions to
+ feature attributes.
```
Comment (1ec5) [inline platform/darwin/src/MGLLight.h]: Note that, other than documentation, this PR hasn’t migrated light properties to expressions yet.

**Comment by 1ec5** [comment]
Certainly, there’s more design discussion about possible conveniences in #8074, with an eye toward including it in the final release as tail work. Note that there’s already a built-in convenience for constant expressions. For example, this line:

```objc
layer.backgroundColor = [NSExpression expressionWithFormat:@"%@", [UIColor redColor]];
```

is equivalent to:

```objc
layer.backgroundColor = [NSExpression expressionForConstantValue:[UIColor redColor]];
```

While the latter isn’t much shorter, it does benefit more from code completion.

For step expressions, the first and third approaches in https://github.com/mapbox/mapbox-gl-native/issues/8074#issuecomment-355383677 could look like:

```objc
layer.backgroundPattern = [NSExpression expressionWithFormat:
                           @"mgl_stepWithMinimum:stops:($zoomLevel, %@, %@)",
                           constantExpression, @{@18: constantExpression}];
layer.backgroundPattern = [NSExpression mgl_stepExpressionWithInput:[NSExpression expressionWithFormat:@"$zoomLevel"]
                                                            minimum:constantExpression
                                                              stops:@{@18: constantExpression}];
```

File: platform/darwin/src/MGLStyleValue.mm [hunk 4]
```diff
@@ -1,8 +1,130 @@
...
+        });
+        return @[@"literal", inputs.lastObject];
+    }
+    if (auto assertExpression = dynamic_cast<const Assertion *>(&mbglExpression)) {
+#warning Implement assertion expressions.
+        NSMutableArray *inputs = [NSMutableArray array];
+        assertExpression->eachChild([&](const Expression &child) {
+            [inputs addObject:MGLJSONObjectFromMBGLExpression(child)];
+        });
+//        type::Type type = assertExpression->getType();
```
Comment (fabian-guerra) [inline platform/darwin/src/MGLStyleValue.mm]: Nit: If this code is not used could you please remove it.

**Comment by 1ec5** [comment]
@fabian-guerra, thank you for taking the time to review this massive PR. 😊

> `MGLLight` documentation already points to `NSExpression` could you please also migrate it's properties to `NSExpression`? so we can get rid of `MGLStyleValue`.

Done. The “Using Style Functions at Runtime” guide still needs to be rewritten and folded into the “Predicates and Expressions” guide.

File: platform/darwin/src/MGLStyleValue.mm (hunk 4 above)
Comment (1ec5) [reply platform/darwin/src/MGLStyleValue.mm]: * [ ] File tail-work ticket about implementing type assertions
* [ ] Remove commented out type assertion code
